src/app/shared/services/courses.service.ts

```diff
@@ -3,16 +3,17 @@ import { Injectable } from '@angular/core';
 import { HttpCoursesService } from '@app/api';
 import { Course } from '@app/models';
 
-import { Observable } from 'rxjs';
+import { Observable, BehaviorSubject } from 'rxjs';
+import { debounceTime, distinctUntilChanged, switchMap } from 'rxjs/operators';
 
 import { v4 as uuidv4 } from 'uuid';
 
 @Injectable({
   providedIn: 'root'
 })
 export class CoursesService {
-
   private paginationSize: number = 5;
+  private searchSubject$ = new BehaviorSubject<string>(null);
```

Comment:
move this subject to search course component. Emit and denounce the value in that component